spring-cloud-gcp-starters/spring-cloud-gcp-starter-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/autoconfig/GcpPubsubAutoConfiguration.java

```diff
@@ -0,0 +1,111 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.autoconfig;
+
+import java.util.concurrent.Executors;
+
+import com.google.api.gax.grpc.ChannelProvider;
+import com.google.api.gax.grpc.ExecutorProvider;
+import com.google.api.gax.grpc.FixedExecutorProvider;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.pubsub.v1.SubscriptionAdminSettings;
+import com.google.cloud.pubsub.v1.TopicAdminSettings;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.cloud.gcp.core.GcpProperties;
+import org.springframework.cloud.gcp.core.autoconfig.GcpContextAutoConfiguration;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
+import org.springframework.cloud.gcp.pubsub.support.DefaultPublisherFactory;
+import org.springframework.cloud.gcp.pubsub.support.DefaultSubscriberFactory;
+import org.springframework.cloud.gcp.pubsub.support.PublisherFactory;
+import org.springframework.cloud.gcp.pubsub.support.SubscriberFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author João André Martins
+ */
+@Configuration
+@AutoConfigureAfter(GcpContextAutoConfiguration.class)
+public class GcpPubsubAutoConfiguration {
+
+	public static final String DEFAULT_SOURCE_NAME = "spring";
+
+	@Value("${spring.cloud.gcp.pubsub.subscriber.executorThreads:4}")
+	private int subscriberExecutorThreads;
+
+	@Value("${spring.cloud.gcp.pubsub.publisher.executorThreads:4}")
+	private int publisherExecutorThreads;
+
+	@Bean
+	public ExecutorProvider publisherExecutorProvider() {
+		return FixedExecutorProvider.create(
+				Executors.newScheduledThreadPool(this.publisherExecutorThreads));
+	}
+
+	@Bean
+	public ExecutorProvider subscriberExecutorProvider() {
+		return FixedExecutorProvider.create(
+				Executors.newScheduledThreadPool(this.subscriberExecutorThreads));
+	}
+
+	@Bean
+	public ChannelProvider subscriberChannelProvider(GoogleCredentials credentials) {
+		return SubscriptionAdminSettings.defaultChannelProviderBuilder()
+				.setCredentialsProvider(() -> credentials)
+				.setClientLibHeader(DEFAULT_SOURCE_NAME,
+						this.getClass().getPackage().getImplementationVersion())
+				.build();
+	}
+
+	@Bean
+	public ChannelProvider publisherChannelProvider(GoogleCredentials credentials) {
```

Comment:
Good question. Because the Google Pubsub API provides distinct ways to get channel providers for both publishers and subscribers, even though I think the final object is the same. Still, I thought it would be a good practice to use both ways, depending on the case.

In any case, I'm preparing another PR on the side where `PublisherFactory` and `SubscriberFactory` will no longer take a `ChannelProvider`. The `ChannelProvider` used to be the way in which the user provided the `GoogleCredentials` for the API calls, but that is no longer the case -- `Publisher` and `Subscriber's` Builders now take a `CredentialsProvider`, and there is nothing else in the `ChannelProvider` left for users to configure that comes to mind.